File: src/connections/usb/UsbDeviceManager.java
package connections.usb;

import java.io.IOException;
import davinci.Device;
import davinci.Master;
import enums.ConnectionMode;
import gui.Home;
import gui.img.ImageIcons;

public class UsbDeviceManager
{
    private UsbDeviceConnection udc;

    private Device usbDevice;
    private boolean backgroundScannerRunning = true;

    public void connect()
        throws IOException
    {
        udc = new UsbDeviceConnection();
        udc.connect();
    }

    public void startBackgroundScanner()
    {
        final Thread t = new Thread(
            new Runnable()
            {
                @Override
                public void run()
                {
                    usbDevice = new Device(
                        ImageIcons.getInstance().getUsbIcon(), new String[] {"USB", "USB", "USB"});

                    int dcCount = 0; // Seconds no UDP signal is detected
                    while (backgroundScannerRunning)
                    {
                        try
                        {
                            Thread.sleep(1000);

                            if (Adb.usbAvailable()) // Will not show a window
                            {
                                dcCount = 0; // Reset disconnect

                                if (!Master.hasConnection(ConnectionMode.USB))
                                {
                                    Master.addConnection(ConnectionMode.USB);
                                    adjustGui(ConnectionMode.USB);
                                }
                            } else
                            {
                                dcCount++;
                                if (dcCount >= 2) // 2 second timeout
                                {
                                    // The USB Device has been disconnected
                                    if (usbDevice != null)
                                    {
                                        if (Master.hasConnection(ConnectionMode.USB))
                                        {
                                            Master.removeConnection(ConnectionMode.USB);
                                            if (Master.hasConnection(ConnectionMode.WIFI))
                                            {
                                                adjustGui(ConnectionMode.WIFI);
                                            } else
                                            {
                                                adjustGui(false);
                                            }
                                        }
                                    }
                                }
                            }
                        } catch (InterruptedException e)
                        {
                            e.printStackTrace();
                        }
                    }
                }
            });

        if (!Adb.isAdbAvailable()) {
            // ADB is unavailable
            Home.getInstance().showErrorPrompt("Error", "ADB not found.");
        } else {
            // Forward ADB
            try
            {
                Adb.startAdb();
            } catch (final IOException e)
            {
                e.printStackTrace();
            }
            t.start();
        }
    }

    public void stopBackgroundScanner()
    {
        backgroundScannerRunning = false;
    }


    // Getters
    public Device getDevice()
    {
        return this.usbDevice;
    }

    private void adjustGui(final ConnectionMode connectionMode)
    {
        Home.getInstance().getDeviceField().show(true);
        if (connectionMode == ConnectionMode.USB)
        {
            getDevice().icon_$eq(ImageIcons.getInstance().getUsbIcon());
            Home.getInstance().getDeviceField().setUi(getDevice());
        } else {
            getDevice().icon_$eq(ImageIcons.getInstance().getWifiIcon());
            Home.getInstance().getDeviceField().setUi(getDevice());
        }
    }

    private void adjustGui(final boolean hidden)
    {
        Home.getInstance().getDeviceField().show(hidden);
    }

    public boolean getBackgroundScannerRunning()
    {
        return this.backgroundScannerRunning;
    }

    public boolean isConnected()
    {
        return udc != null && udc.getConnection() != null && udc.getConnection().isConnected();
    }
}

File: src/main/scala/connections/usb/UsbDeviceManager.scala
package connections.usb

import java.io.IOException

import connections.{BaseDeviceManager, ConnectionManager}
import slide.Device
import enums.ConnectionMode
import gui.ImageIcons

class UsbDeviceManager extends BaseDeviceManager {

    private var udc: UsbDeviceConnection = null
    private var backgroundScannerRunning: Boolean = true

    @throws(classOf[IOException])
    def connect(ip: String): Unit = {
        udc = new UsbDeviceConnection(ip) {
            override def onClientOutOfDate(): Unit = {
                throwError("The client is out of date. Please upgrade it.")
            }
        }
        udc.connect()
    }

    def startBackgroundScanner(): Unit = {
        val t: Thread = new Thread(new Runnable {
            def run() {
                device = new Device(ImageIcons.usbIcon, Array[String]("USB", "USB", "USB"))
                var dcCount: Int = 0
                while (backgroundScannerRunning) {
                    Thread.sleep(1000)
                    if (Adb.usbAvailable) {
                        dcCount = 0
                        if (!ConnectionManager.hasConnection(ConnectionMode.USB)) {
                            onUsbConnectionAdded()
                        }
                    }
                    else {
                        dcCount += 1
                        if (dcCount >= 2 && device != null && ConnectionManager.hasConnection(ConnectionMode.USB)) {
                            onUsbConnectionRemoved()
                        }
                    }
                }
            }
        })
        if (!Adb.isAdbAvailable) {
            throwError("Adb not found")
        } else {
            t.start()
        }
    }

    override def throwError(message: String): Unit = {}

    def stopBackgroundScanner(): Unit = backgroundScannerRunning = false
}
